File: doc-examples/nitrite-example-java/src/test/java/example/PersonRepositoryTest.java
package example;

import io.micronaut.data.model.Pageable;
import io.micronaut.data.model.Sort;
import io.micronaut.data.repository.jpa.criteria.PredicateSpecification;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import jakarta.inject.Inject;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Optional;

import static io.micronaut.data.repository.jpa.criteria.PredicateSpecification.where;
import static org.junit.jupiter.api.Assertions.*;

@MicronautTest(transactional = false)
class PersonRepositoryTest {

    @Inject
    PersonRepository personRepository;

    @AfterEach
    void cleanup() {
        personRepository.deleteAll();
    }

    @Test
    void testSaveAndFind() {
        Person person = new Person("Denis", 30);
        person.setInterests(List.of("Java", "Micronaut"));
        personRepository.save(person);

        Optional<Person> found = personRepository.findOne(PersonRepository.Specifications.nameEquals("Denis"));
        assertTrue(found.isPresent());
        assertEquals("Denis", found.get().getName());
        assertEquals(30, found.get().getAge());
        assertTrue(found.get().getInterests().contains("Java"));
    }

    @Test
    void testCountSpecifications() {
        personRepository.save(new Person("Denis", 30));
        personRepository.save(new Person("Josh", 25));
        personRepository.save(new Person("John", 35));

        long countAgeLess40 = personRepository.count(PersonRepository.Specifications.ageIsLessThan(40));
        assertEquals(3, countAgeLess40);

        long countAgeLess20 = personRepository.count(PersonRepository.Specifications.ageIsLessThan(20));
        assertEquals(0, countAgeLess20);
    }

    @Test
    void testFindAllWithSpecifications() {
        personRepository.save(new Person("Denis", 30));
        personRepository.save(new Person("Josh", 25));
        personRepository.save(new Person("John", 35));

        PredicateSpecification<Person> spec = PersonRepository.Specifications.nameEquals("Denis")
            .or(PersonRepository.Specifications.nameEquals("Josh"));
        List<Person> people = personRepository.findAll(spec);
        assertEquals(2, people.size());
    }

    @Test
    void testSorting() {
        personRepository.save(new Person("Charlie", 30));
        personRepository.save(new Person("Alice", 25));
        personRepository.save(new Person("Bob", 35));

        List<Person> sorted = personRepository.findAll(Sort.of(Sort.Order.asc("name")));
        assertEquals(3, sorted.size());
        assertEquals("Alice", sorted.get(0).getName());
        assertEquals("Bob", sorted.get(1).getName());
        assertEquals("Charlie", sorted.get(2).getName());
    }

    @Test
    void testPagination() {
        for (int i = 1; i <= 10; i++) {
            personRepository.save(new Person("Person" + i, 20 + i));
        }

        Pageable pageable = Pageable.from(0, 5, Sort.of(Sort.Order.asc("name")));
        var page = personRepository.findAll(pageable);

        assertEquals(10, page.getTotalSize());
        assertEquals(2, page.getTotalPages());
        assertEquals(5, page.getSize());
    }

    @Test
    void testSortingAndPaginationCombined() {
        personRepository.save(new Person("Denis", 30));
        personRepository.save(new Person("Josh", 25));
        personRepository.save(new Person("John", 35));
        personRepository.save(new Person("Alice", 20));

        Pageable pageable = Pageable.from(0, 2, Sort.of(Sort.Order.asc("name")));
        var page = personRepository.findAll(pageable);

        assertEquals(4, page.getTotalSize());
        assertEquals(2, page.getSize());
        assertEquals("Alice", page.getContent().get(0).getName());
        assertEquals("Denis", page.getContent().get(1).getName());
    }

    @Test
    void testUpdateAll() {
        personRepository.save(new Person("Denis", 30));
        personRepository.save(new Person("Denis Jr", 25));

        var updateSpec = PersonRepository.Specifications.setNewName("Steven");
        long recordsUpdated = personRepository.updateAll(updateSpec.where(PersonRepository.Specifications.nameEquals("Denis")));
        assertEquals(1, recordsUpdated);

        Optional<Person> updated = personRepository.findOne(PersonRepository.Specifications.nameEquals("Steven"));
        assertTrue(updated.isPresent());
        assertEquals("Steven", updated.get().getName());
    }

    @Test
    void testDeleteAll() {
        personRepository.save(new Person("Denis", 30));
        personRepository.save(new Person("Josh", 25));
        personRepository.save(new Person("John", 35));

        long recordsDeleted = personRepository.deleteAll(where(PersonRepository.Specifications.nameEquals("Denis")));
        assertEquals(1, recordsDeleted);

        assertEquals(2, personRepository.count());
    }
}

File: doc-examples/nitrite-example-kotlin/src/test/kotlin/example/PersonRepositoryTest.kt
package example

import io.micronaut.data.model.Pageable
import io.micronaut.data.model.Sort
import io.micronaut.test.extensions.junit5.annotation.MicronautTest
import jakarta.inject.Inject
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertTrue

@MicronautTest(transactional = false)
class PersonRepositoryTest {

    @Inject
    lateinit var personRepository: PersonRepository

    @AfterEach
    fun cleanup() {
        personRepository.deleteAll()
    }

    @Test
    fun testSaveAndFind() {
        val person = Person("Denis", 30)
        person.interests = mutableListOf("Java", "Micronaut")
        personRepository.save(person)

        val found = personRepository.findOne(PersonRepository.Specifications.nameEquals("Denis")).orElse(null)
        assertNotNull(found)
        assertEquals("Denis", found!!.name)
        assertEquals(30, found.age)
        assertTrue(found.interests?.contains("Java") ?: false)
    }

    @Test
    fun testCountSpecifications() {
        personRepository.save(Person("Denis", 30))
        personRepository.save(Person("Josh", 25))
        personRepository.save(Person("John", 35))

        val countAgeLess40 = personRepository.count(PersonRepository.Specifications.ageIsLessThan(40))
        assertEquals(3, countAgeLess40)

        val countAgeLess20 = personRepository.count(PersonRepository.Specifications.ageIsLessThan(20))
        assertEquals(0, countAgeLess20)
    }

    @Test
    fun testFindAllWithSpecifications() {
        personRepository.save(Person("Denis", 30))
        personRepository.save(Person("Josh", 25))
        personRepository.save(Person("John", 35))

        val spec = PersonRepository.Specifications.nameEquals("Denis")
            .or(PersonRepository.Specifications.nameEquals("Josh"))
        val people = personRepository.findAll(spec)
        assertEquals(2, people.size)
    }

    @Test
    fun testSorting() {
        personRepository.save(Person("Charlie", 30))
        personRepository.save(Person("Alice", 25))
        personRepository.save(Person("Bob", 35))

        val sorted = personRepository.findAll(Sort.of(Sort.Order.asc("name")))
        assertEquals(3, sorted.size)
        assertEquals("Alice", sorted[0].name)
        assertEquals("Bob", sorted[1].name)
        assertEquals("Charlie", sorted[2].name)
    }

    @Test
    fun testPagination() {
        for (i in 1..10) {
            personRepository.save(Person("Person$i", 20 + i))
        }

        val pageable = Pageable.from(0, 5, Sort.of(Sort.Order.asc("name")))
        val page = personRepository.findAll(pageable)

        assertEquals(10, page.totalSize)
        assertEquals(2, page.totalPages)
        assertEquals(5, page.size)
    }

    @Test
    fun testSortingAndPaginationCombined() {
        personRepository.save(Person("Denis", 30))
        personRepository.save(Person("Josh", 25))
        personRepository.save(Person("John", 35))
        personRepository.save(Person("Alice", 20))

        val pageable = Pageable.from(0, 2, Sort.of(Sort.Order.asc("name")))
        val page = personRepository.findAll(pageable)

        assertEquals(4, page.totalSize)
        assertEquals(2, page.size)
        assertEquals("Alice", page.content[0].name)
        assertEquals("Denis", page.content[1].name)
    }

    @Test
    fun testUpdateAll() {
        personRepository.save(Person("Denis", 30))
        personRepository.save(Person("Denis Jr", 25))

        val updateSpec = PersonRepository.Specifications.setNewName("Steven")
        val recordsUpdated = personRepository.updateAll(updateSpec.where(PersonRepository.Specifications.nameEquals("Denis")))
        assertEquals(1, recordsUpdated)

        val updated = personRepository.findOne(PersonRepository.Specifications.nameEquals("Steven")).orElse(null)
        assertNotNull(updated)
        assertEquals("Steven", updated!!.name)
    }

    @Test
    fun testDeleteAll() {
        personRepository.save(Person("Denis", 30))
        personRepository.save(Person("Josh", 25))
        personRepository.save(Person("John", 35))

        val spec = PersonRepository.Specifications.nameEquals("Denis")
        val recordsDeleted = personRepository.deleteAll(spec)
        assertEquals(1, recordsDeleted)

        assertEquals(2, personRepository.count())
    }
}
